File: TRViS.UITests/Tests/NavigationTests.cs
using TRViS.UITests.Pages;

namespace TRViS.UITests.Tests;

/// <summary>
/// Navigation tests require Firebase consent to have been accepted first.
/// Each test accepts Firebase consent at the beginning.
/// </summary>
[TestFixture]
[Infrastructure.RetryAllTests(2)] // see AppLaunchTests for rationale
public class NavigationTests : BaseUITest
{
	// Share one Appium session across all tests in this fixture (iOS only).
	// See BaseUITest.ShareSessionAcrossTestsInFixture for details.
	protected override bool ShareSessionAcrossTestsInFixture => true;

	private AppShellPage _shell = null!;

	[SetUp]
	public override void SetUp()
	{
		base.SetUp();

		// Navigate to StartHome at the top of each test so OpenFlyout starts
		// from a flyout-rooted page. The earlier "OpenFlyout works from any
		// flyout-capable page" assumption breaks on Android in the
		// shared-session run: the previous fixture (HorizontalTimetableTests)
		// lands the app on DTAC's VerticalView tab which has orientation
		// locked to Landscape, and on Android the OpenFlyout probes (DTAC
		// MenuButton click or "Open navigation drawer") don't open the
		// NavigationView under that lock (CI run 25727806170 / 25729263553).
		// NavigateToHome uses the DTAC.TestNavigateHomeButton seam to GoToAsync
		// past the broken flyout path, and the seam's OnDisappearing also
		// unlocks the orientation. From StartHome OpenFlyout works fine on
		// every platform.
		var startHome = new StartHomePageObject(Driver);
		if (!startHome.PollDisplayed(AutomationIds.StartHome.Title, timeoutSeconds: 3))
		{
			new AppShellPage(Driver).NavigateToHome();
			startHome = new StartHomePageObject(Driver);
		}
		startHome.AcceptPrivacyPolicyIfNeeded();
		startHome.ClearLoaderForTesting();

		_shell = new AppShellPage(Driver);
	}

	/// <summary>
	/// TPL is no longer a flyout entry — it is reached via the StartHome
	/// footer "Third Party Licenses" link, which pushes it as a modal with
	/// an in-page close (X) icon. Opens it, asserts it renders, then closes
	/// via the icon and asserts we return to StartHome.
	/// </summary>
	[Test]
	public void Footer_OpenAndCloseThirdPartyLicenses()
	{
		var startHome = new StartHomePageObject(Driver);

		var page = startHome.OpenThirdPartyLicenses();
		Assert.That(page.IsDisplayed(), Is.True,
			"ThirdPartyLicenses modal should be displayed after tapping the footer link.");

		startHome = page.CloseModal();
		Assert.That(startHome.IsDisplayed(), Is.True,
			"StartHomePage should be displayed again after closing the TPL modal.");
	}

	[Test]
	public void Flyout_NavigateToSettings()
	{
		var page = _shell.NavigateToSettings();
		Assert.That(page.IsDisplayed(), Is.True,
			"Settings (EasterEgg) page should be displayed after navigation.");
	}

	[Test]
	public void Flyout_NavigateToDTAC()
	{
		// On Android the DTAC FlyoutItem is removed (MAUI #16927 mitigation):
		// DTAC is a relative route reachable only after a Work is committed.
		// Flyout-based navigation to DTAC no longer exists on this platform.
		if (IsAndroid)
			Assert.Ignore("Android: DTAC flyout item removed (MAUI #16927); DTAC reachable only via relative route after Work commit.");

		var page = _shell.NavigateToDTAC();
		Assert.That(page.IsDisplayed(), Is.True,
			"DTACViewHost page should be displayed after navigation.");
	}

	/// <summary>
	/// Regression for #240: ViewHost is a cached ShellContent (DataTemplate),
	/// so the same instance is reused on every flyout navigation. Earlier code
	/// disposed the presenter on Unloaded, permanently severing its
	/// LocationService.TimeChanged subscription — the second visit then showed
	/// a frozen clock. Asserts the clock actually changes on the second visit.
	///
	/// Reads the time via DTAC.TestTimeSeam (a UI_TEST-only Label that mirrors
	/// the presenter's TimeLabelText with a sentinel prefix). Cannot use the
	/// real DTAC.TimeLabel: iOS doesn't surface MAUI Labels in the
	/// accessibility tree when their text is empty, and the AppBar's TimeLabel
	/// is hidden on narrow phones in portrait by an internal width threshold.
	/// </summary>
	[Test]
	public void DTAC_ReopenAfterNavigateAway_ClockKeepsTicking()
	{
		// This regression tests the *cached* ShellContent model (#240): the same
		// ViewHost instance is reused across visits. On Android (MAUI #16927
		// mitigation), ViewHost is a fresh route-created instance per visit — the
		// cached-instance failure mode does not apply. Android DTAC tear-down
		// coverage is provided by Dtac_TearDownRepro_Tests instead.
		if (IsAndroid)
			Assert.Ignore("Android: ViewHost is route-created (fresh per visit); cached-ShellContent regression (#240) does not apply.");

		var dtac = _shell.NavigateToDTAC();
		Assert.That(dtac.IsDisplayed(), Is.True,
			"DTAC should be visible on the first visit.");

		_shell.NavigateToHome();

		dtac = _shell.NavigateToDTAC();
		Assert.That(dtac.IsDisplayed(), Is.True,
			"DTAC should be visible on the second visit.");

		// LocationService raises TimeChanged once per second (Task.Delay(100)
		// polling that fires only when GetCurrentTimeSeconds() advances).
		// Sleep > 1 s between reads so a live clock must have ticked at least
		// once even if the first read landed right after a tick.
		string firstReading = dtac.ReadTimeViaSeam();
		Thread.Sleep(1500);
		string secondReading = dtac.ReadTimeViaSeam();

		Assert.That(secondReading, Is.Not.EqualTo(firstReading),
			$"Clock must keep updating on the second DTAC visit (#240). " +
			$"First='{firstReading}', Second='{secondReading}'.");
	}

	/// <summary>
	/// Regression for #240 (title half): the same disposed-presenter chain
	/// also unhooked AppViewModel.PropertyChanged, so a SelectedWork change
	/// that happened *while DTAC was hidden* never updated the AppBar title.
	/// Repro: open DTAC empty (no Work) → home → load sample + commit Work →
	/// DTAC. With the bug, AppBar Title stays empty on the second visit.
	///
	/// Reads via DTAC.TestTitleSeam (UI_TEST-only mirror Label) for the same
	/// iOS-accessibility reason as the clock test above.
	/// </summary>
	[Test]
	public void DTAC_ReopenAfterWorkSelected_TitleUpdated()
	{
		// Same rationale as DTAC_ReopenAfterNavigateAway_ClockKeepsTicking: this
		// test exercises the cached ShellContent model (#240). On Android, ViewHost
		// is route-created (fresh per visit) so the presenter-disposal failure mode
		// cannot occur. Skip on Android; see Dtac_TearDownRepro_Tests for Android
		// DTAC coverage.
		if (IsAndroid)
			Assert.Ignore("Android: ViewHost is route-created (fresh per visit); cached-ShellContent regression (#240) does not apply.");

		// Clear any loader/SelectedWork left by a prior test in this
		// fixture's shared Appium session — base SetUp only navigates to
		// StartHome and accepts the privacy policy. Without this, a
		// non-empty AppBar title carried over from an earlier test would
		// let the later `Is.Not.EqualTo(initialTitle)` assertion pass
		// trivially even with the bug present.
		var startHome = new StartHomePageObject(Driver);
		startHome.ClearLoaderForTesting();

		// First DTAC visit with no Work selected — primes the broken
		// codepath by triggering Unloaded → Dispose on the way back.
		var dtac = _shell.NavigateToDTAC();
		Assert.That(dtac.IsDisplayed(), Is.True);

		// Capture the post-clear title baseline. With no SelectedWork the
		// presenter sets TitleText to "", so the seam reads its sentinel
		// prefix only (stripped to "" by the page object). Asserting the
		// later title differs from this baseline catches the bug even on
		// platforms where an empty TitleText still surfaces non-empty
		// fallback text in the seam.
		string initialTitle = dtac.ReadTitleViaSeam();

		startHome = _shell.NavigateToHome();
		// Demo data load + auto-open commits SelectedWork AND navigates to
		// DTAC in one shot. Both side-effects happen *after* the dead
		// presenter from the first visit would have missed them.
		startHome.LoadSample();
		startHome.WaitForElement(AutomationIds.StartHome.WorkGroupList);
		dtac = startHome.AutoOpenForTesting();
		Assert.That(dtac.IsDisplayed(), Is.True,
			"DTAC should be visible after AutoOpenForTesting.");

		// Title set goes through MainThread.BeginInvokeOnMainThread; give
		// the dispatcher a beat to flush before reading.
		Thread.Sleep(500);
		string title = dtac.ReadTitleViaSeam();
		Assert.That(title, Is.Not.Empty,
			$"AppBar Title must reflect the committed Work on the second " +
			$"DTAC visit (#240). Got='{title}'.");
		Assert.That(title, Is.Not.EqualTo(initialTitle),
			$"AppBar Title must change after a Work is selected (#240). " +
			$"Initial='{initialTitle}', After='{title}'.");

		// Restore Start mode for any subsequent test in this fixture's
		// shared session. Mirrors DTACTimetableTests.SetUp's assumption.
		_shell.NavigateToHome().ClearLoaderForTesting();
	}
}

File: TRViS.UITests.Apple/Sources/Shared/NavigationTests.swift
// NavigationTests.swift
// Mirrors TRViS.UITests/Tests/NavigationTests.cs (5 tests).
//
// Each test uses a fresh per-test app launch (BaseUITestCase.setUp) and
// accepts the privacy policy if needed at the start of every test.
// No shared session is used; the C# shared-session setup recovery blocks
// are dropped — per-test launch makes them unnecessary.

import XCTest

final class NavigationTests: BaseUITestCase {

    // Per-test setUp: after super.setUpWithError() the app is freshly launched.
    // Accept privacy if the banner is visible (first test in a run), then
    // construct the shell page object.
    override func setUpWithError() throws {
        try super.setUpWithError()
        let startHome = StartHomePageObject(app: app, base: self)
        XCTAssertTrue(
            startHome.isDisplayed(),
            "StartHomePage should be displayed on launch (NavigationTests setUp)."
        )
        startHome.acceptPrivacyPolicyIfNeeded()
    }

    // MARK: — Test: Footer_OpenAndCloseThirdPartyLicenses

    /// Mirrors C# NavigationTests.Footer_OpenAndCloseThirdPartyLicenses.
    ///
    /// TPL is reached via the StartHome footer "Third Party Licenses" link
    /// (flyout entry was removed). Opens it, asserts it renders, closes via
    /// the modal X icon, asserts we return to StartHome.
    func testFooter_OpenAndCloseThirdPartyLicenses() throws {
        let startHome = StartHomePageObject(app: app, base: self)

        let tpl = startHome.openThirdPartyLicenses()
        XCTAssertTrue(
            tpl.isDisplayed(),
            "ThirdPartyLicenses modal should be displayed after tapping the footer link."
        )

        let startHomeAgain = tpl.closeModal()
        XCTAssertTrue(
            startHomeAgain.isDisplayed(),
            "StartHomePage should be displayed again after closing the TPL modal."
        )
    }

    // MARK: — Test: Flyout_NavigateToSettings

    /// Mirrors C# NavigationTests.Flyout_NavigateToSettings.
    func testFlyout_NavigateToSettings() throws {
        let shell = AppShellPageObject(app: app, base: self)
        let page = shell.navigateToSettings()
        XCTAssertTrue(
            page.isDisplayed(),
            "Settings (EasterEgg) page should be displayed after navigation."
        )
    }

    // MARK: — Test: Flyout_NavigateToDTAC

    /// Mirrors C# NavigationTests.Flyout_NavigateToDTAC.
    func testFlyout_NavigateToDTAC() throws {
        let shell = AppShellPageObject(app: app, base: self)
        let page = shell.navigateToDTAC()
        XCTAssertTrue(
            page.isDisplayed(),
            "DTACViewHost page should be displayed after navigation."
        )
    }

    // MARK: — Test: DTAC_ReopenAfterNavigateAway_ClockKeepsTicking

    /// Mirrors C# NavigationTests.DTAC_ReopenAfterNavigateAway_ClockKeepsTicking.
    ///
    /// Regression for #240: checks that the clock label keeps updating on the
    /// second DTAC visit. Reads via DTAC.TestTimeSeam so narrow-phone / empty-text
    /// iOS accessibility-tree limitations don't affect the result.
    func testDTAC_ReopenAfterNavigateAway_ClockKeepsTicking() throws {
        let shell = AppShellPageObject(app: app, base: self)

        let dtac = shell.navigateToDTAC()
        XCTAssertTrue(dtac.isDisplayed(), "DTAC should be visible on the first visit.")

        _ = shell.navigateToHome()

        let dtac2 = shell.navigateToDTAC()
        XCTAssertTrue(dtac2.isDisplayed(), "DTAC should be visible on the second visit.")

        let firstReading = dtac2.readTimeViaSeam()
        // LocationService ticks every ~1 s; sleep 1.5 s to guarantee at least
        // one tick even if the first read landed right after a tick.
        Thread.sleep(forTimeInterval: 1.5)
        let secondReading = dtac2.readTimeViaSeam()

        XCTAssertNotEqual(
            secondReading, firstReading,
            "Clock must keep updating on the second DTAC visit (#240). " +
            "First='\(firstReading)', Second='\(secondReading)'."
        )
    }

    // MARK: — Test: DTAC_ReopenAfterWorkSelected_TitleUpdated

    /// Mirrors C# NavigationTests.DTAC_ReopenAfterWorkSelected_TitleUpdated.
    ///
    /// Regression for #240 (title half): asserts AppBar title updates when a Work
    /// is selected after the initial empty DTAC visit that primes the broken codepath.
    func testDTAC_ReopenAfterWorkSelected_TitleUpdated() throws {
        let startHome = StartHomePageObject(app: app, base: self)
        let shell = AppShellPageObject(app: app, base: self)

        // Clear any leftover loader state (belt-and-suspenders on per-test launch,
        // but mirrors the C# fixture's intent explicitly).
        startHome.clearLoaderForTesting()

        // First DTAC visit with no Work selected — primes the broken codepath
        // by triggering Unloaded → Dispose on the way back.
        let dtac = shell.navigateToDTAC()
        XCTAssertTrue(dtac.isDisplayed(), "DTAC should be visible on the first visit.")

        let initialTitle = dtac.readTitleViaSeam()

        _ = shell.navigateToHome()

        // Load demo data and auto-open → DTAC with the first Work committed.
        startHome.loadSample()
        _ = startHome.waitForWorkGroupList(timeout: 30)

        let dtac2 = startHome.autoOpenForTesting()
        XCTAssertTrue(
            dtac2.isDisplayed(),
            "DTAC should be visible after AutoOpenForTesting."
        )

        // Title set goes through MainThread.BeginInvokeOnMainThread; give the
        // dispatcher a beat to flush before reading.
        Thread.sleep(forTimeInterval: 0.5)
        let title = dtac2.readTitleViaSeam()

        XCTAssertFalse(
            title.isEmpty,
            "AppBar Title must reflect the committed Work on the second DTAC visit (#240). " +
            "Got='\(title)'."
        )
        XCTAssertNotEqual(
            title, initialTitle,
            "AppBar Title must change after a Work is selected (#240). " +
            "Initial='\(initialTitle)', After='\(title)'."
        )
    }
}
